core/trie/proof.go

```diff
@@ -0,0 +1,80 @@
+package trie
+
+import (
+	"errors"
+
+	"github.com/NethermindEth/juno/core/crypto"
+	"github.com/NethermindEth/juno/core/felt"
+)
+
+type Membership int
+
+const (
+	Member Membership = iota
+	NonMember
+)
+
+type ProofNode struct {
+	LeftHash  *felt.Felt
+	RightHash *felt.Felt
```

Comment:
![image](https://github.com/NethermindEth/juno/assets/775020/040bc744-bcdc-4c4d-a479-72e7273cb318)

to better understand what's going on I attached image of a binary tree 😄 
so let's say we need to proof the presence of node 9 in the trie, we will return ProofNodes from root to the leaf (9):
[50, 17, 12, 9]. Question is: since we return just one branch of a trie why do we need to calculate rightHash (in this example) at all?

Reply:
So we actually only return the proofNodes of [50,17,12] which are the hashes of the left and right child.
To verify these root nodes, we need to also supply the root hash, the leaf hash, and the path down to the leaf hash.

We effectively check that:
- rootHash = H(hash_17,hash_72) 
- hash_17 = H(hash_12, hash_23)
- hash_12 = H(hash_9, hash_14)
- hash_9 = leafHash
where the path was used to select the above nodes. If at any point this fails (eg hash9!=leafHash) then we know the path doesn't lead from the root to the leaf along the provided proofNodes.

Reply:
Note: You need the edge to the branch also. https://github.com/starknet-io/starknet-p2p-specs/blob/main/p2p/proto/snapshot.proto#L6